dm/pkg/checker/table_structure.go

```diff
@@ -79,68 +95,128 @@ func (c *TablesChecker) Check(ctx context.Context) *Result {
 		Name:  c.Name(),
 		Desc:  "check compatibility of table structure",
 		State: StateSuccess,
-		Extra: fmt.Sprintf("address of db instance - %s:%d", c.dbinfo.Host, c.dbinfo.Port),
 	}
 
+	startTime := time.Now()
 	var (
-		err        error
-		options    = make(map[string][]*incompatibilityOption)
-		statements = make(map[string]string)
+		inCh          = make(chan *checkItem, maxThreadNum)
+		checkWg       sync.WaitGroup
+		reMu          sync.Mutex
+		firstInstance string
 	)
-	for schema, tables := range c.tables {
-		if len(tables) == 0 {
-			tables, err = dbutil.GetTables(ctx, c.db, schema)
-			if err != nil {
-				markCheckError(r, err)
-				return r
-			}
-		}
 
-		for _, table := range tables {
-			tableName := dbutil.TableName(schema, table)
-			statement, err := dbutil.GetCreateTableSQL(ctx, c.db, schema, table)
-			if err != nil {
-				// continue if table was deleted when checking
-				if isMySQLError(err, mysql.ErrNoSuchTable) {
-					continue
+	for instance := range c.tableMap {
+		firstInstance = instance
+		break
+	}
+
+	checkCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	checkFunc := func() {
+		defer checkWg.Done()
+		instance := firstInstance
+		p, err := dbutil.GetParserForDB(ctx, c.dbs[instance])
+		if err != nil {
+			reMu.Lock()
+			markCheckError(r, err)
+			reMu.Unlock()
+			return
+		}
+		for {
+			select {
+			case <-checkCtx.Done():
+				return
+			case checkItem, ok := <-inCh:
+				if !ok {
+					return
+				}
+				table := checkItem.table
+				if instance != checkItem.sourceID {
+					instance = checkItem.sourceID
+					p, err = dbutil.GetParserForDB(ctx, c.dbs[instance])
```

Comment:
can we share global parser with checker?,(not sure `ParseOneStmt` is thread safe), if so, we can move the operation of change parser for db out of for loop

Reply:
can't, `ParseOneStmt` is not thread safe.